rust/headless-client/src/imp_linux.rs

```diff
@@ -30,14 +30,36 @@ const ROOT_USER: u32 = 0;
 /// for security, so we're following their lead. `/run` and `/var/run` are symlinked
 /// on some systems, `/run` should be the newer version.
 const SOCK_PATH: &str = "/run/firezone-client.sock";
+const TOKEN_ENV_KEY: &str = "FIREZONE_TOKEN";
 
 pub fn default_token_path() -> PathBuf {
     PathBuf::from("/etc")
         .join(connlib_shared::BUNDLE_ID)
         .join("token")
 }
 
-pub async fn run() -> Result<()> {
+pub fn run() -> Result<()> {
+    // Modifying the environment of a running process is unsafe. If any other
+    // thread is reading or writing the environment, something bad can happen.
+    // So `run` must take over as early as possible during startup, and
+    // take the token env var before any other threads spawn.
+    let token = std::env::var(TOKEN_ENV_KEY).ok().map(SecretString::from);
+    // Docs indicate that `remove_var` should actually be marked unsafe
+    // SAFETY: We haven't spawned any other threads, this code should be the first
+    // thing to run after entering `main`. So nobody else is reading the environment.
+    #[allow(unused_unsafe)]
+    unsafe {
+        // This removes the token from the environment per <https://security.stackexchange.com/a/271285>. We run as root so it may not do anything besides defense-in-depth.
+        std::env::remove_var(TOKEN_ENV_KEY);
+    }
+    assert!(std::env::var(TOKEN_ENV_KEY).is_err());
+
+    let rt = tokio::runtime::Runtime::new()?;
```

Comment:
we can go with a single worker thread like we do in the other client applications